### PR TITLE
Upgrade statinamic to latest (^0.8)

### LINKED ISSUE #497: Add back custom 404
https://github.com/putaindecode/putaindecode.fr/pull/495#issuecomment-153080020

(This issue will make more sense when #495 will be merged :D)

### LINKED ISSUE #512: Checksum issue in / and /posts/
Not the same issue as the previous in #501 
Maybe related to collection rendering.

### PR DESCRIPTION
**Pull request by MoOx**
Closes #497
Closes #512 

Also bring
- markdown hot loading
- devServer should be faster
- devServer pre-rendering (handy to debug checksum issues)
- devServer 404
- page description as meta (excerpt of 140 chars, added in html meta/og/twitter)  (that can be used in post listing too in the future).

### CODE REVIEW

File: scripts/index-client.js [hunk 1]
```diff
@@ -1,23 +1,35 @@
...
 statinamicClient({
+  layouts,
   metadata,
   routes,
   store,
 })
+
+// md files → JSON && generate collection + hot loading for dev
+let mdContext = require.context("../content", true, /\.md$/)
+mdContext.keys().forEach(mdContext)
```
Comment (Nyalab) [inline scripts/index-client.js]: wtf?

File: scripts/webpack.config.babel.js [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+import pkg from "../package.json"
+import config from "./config.js"
+
+export default {
+  module: {
+    loaders: [
+      { // statinamic requirement
+        test: /\.md$/,
+        loader: "statinamic/lib/md-collection-loader" +
+          `?${ JSON.stringify({
```
Comment (Nyalab) [inline scripts/webpack.config.babel.js]: move backtick on the line above

File: scripts/webpack.config.babel.js [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+              comments: true,
+            },
+          }) }`,
+      },
+      {
+        test: /styles\.css$/,
+        loader: ExtractTextPlugin.extract(
+          "style-loader",
+          [
+            "css-loader" + "?" + [
```
Comment (Nyalab) [inline scripts/webpack.config.babel.js]: same

File: web_modules/GoogleAnalyticsTracker/index.js [hunk 4]
```diff
@@ -4,16 +4,31 @@ import { PropTypes } from "react"
 import ga from "react-google-analytics"
 const GoogleAnalyticsInitiailizer = ga.Initializer
 
-class GoogleAnalyticsTracker extends Component {
+const isProduction = process.env.NODE_ENV === "production"
+const isClient = typeof window !== "undefined"
+
+export default class GoogleAnalyticsTracker extends Component {
```
Comment (Nyalab) [inline web_modules/GoogleAnalyticsTracker/index.js]: why did you change the way it was declared?

propTypes+contextTypes outside of the component + export at the end of the file is way clearer imo

File: web_modules/layouts/Page/index.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
+import React, { PropTypes } from "react"
+import Helmet from "react-helmet"
+
+const Page = ({ head, body, __url }, context) => {
```
Comment (Nyalab) [inline web_modules/layouts/Page/index.js]:
```js
const Page = ({ head, body, __url }, { metadata }) => ({
  <div>
    ...
  </div>
})
```

**Comment by Nyalab** [comment]
apart from JSX coding style that i will not discuss about right now (so ugly), good work :clap: just a few things to talk about

File: scripts/index-client.js (hunk 1 above)
Comment (MoOx) [inline scripts/index-client.js]: try removing the line, and the all things just don't work. require.context is so weird.

File: scripts/webpack.config.babel.js (hunk 2 above)
Comment (MoOx) [inline scripts/webpack.config.babel.js]: I will do that later: this entire json.stringshit will be (re)moved anyway in a future statinamic upgrade.

File: scripts/webpack.config.babel.js (hunk 3 above)
Comment (MoOx) [inline scripts/webpack.config.babel.js]: same what? Show me how you will indent this all block.

File: web_modules/GoogleAnalyticsTracker/index.js (hunk 4 above)
Comment (MoOx) [inline web_modules/GoogleAnalyticsTracker/index.js]: Sorry I just copy pasted this from original statinamic snippet. Will revert this.

File: scripts/index-client.js (hunk 1 above)
Comment (Nyalab) [inline scripts/index-client.js]: this is really weird, but ok

File: scripts/webpack.config.babel.js (hunk 2 above)
Comment (Nyalab) [inline scripts/webpack.config.babel.js]: ok

File: scripts/webpack.config.babel.js (hunk 3 above)
Comment (Nyalab) [inline scripts/webpack.config.babel.js]: nevermind

File: web_modules/layouts/Page/index.js (hunk 5 above)
Comment (MoOx) [inline web_modules/layouts/Page/index.js]: ? why do you want to return an object? Not sure to follow.

**Comment by MoOx** [comment]
Who want to review this? This will fixes some issues and bring 404 pages back.

File: web_modules/layouts/Page/index.js (hunk 5 above)
Comment (Nyalab) [inline web_modules/layouts/Page/index.js]: sorry, i was typing too fast, just 
```js
```js
const Page = ({ head, body, __url }, { metadata }) => (
  <div>
    ...
  </div>
)
```

Comment (MoOx) [inline web_modules/layouts/Page/index.js]: there is a variable definition before the return.

Comment (Nyalab) [inline web_modules/layouts/Page/index.js]: no, look at the second parameter, i'm destructuring it in the function signature

Comment (MoOx) [inline web_modules/layouts/Page/index.js]: ahah ok. I am currently considering using less destructuring as I find it hide some code relation but I will fix that ok.

Comment (MoOx) [inline web_modules/layouts/Page/index.js]: fixed.

**Comment by MoOx** [comment]
Build now require node 5 (for npm3). We can add compat to npm2 but I am lazy.

**Comment by Nyalab** [comment]
:+1: :clap: it has been 3 days, you can merge it if you want, i support you ;)

**Comment by magsout** [comment]
OOps, sorry. I trust you (easy review, no ?;) ), so GO ON
